### PR TITLE
Change Paper - content updates

### LINKED ISSUE #66: Change paper publication date should be in the format "dd MMM yyyy"
# Overview

Currently, the Mauro produced change papers output the publication date like this:

image

However, the publication date should actually be in "dd MMM yyyy" format, like the original change paper:

image

 # Related

- NHSDigital/DataDictionaryPublication#498

### LINKED ISSUE #71: "Summary of Changes" for change papers of in-flight branches are not matching originals
# Background

This has come about due to comparisons between Borland Together change papers and Mauro generated change papers.

Originals are found in [Borland Together](https://systemccom.sharepoint.com/:f:/s/Mauro/EtIpFN2vnmhHh-EYh3N3IBoB2Da4GbdGUzEChvjiZKgpKA?e=cDxiJR)

Mauro versions are found in [Change Papers](https://systemccom.sharepoint.com/:f:/s/Mauro/Ep_lnVsG5CdIn5sCZvv6miIB_U1zbSoQ7dotYLBlxONKsA?e=anU9HS)

Test reports with more detail can be found in: [In-flight migration testing](https://systemccom.sharepoint.com/:f:/s/Mauro/Ek_bv_Mnng5ErU34rABQstsBPe8Sr6z543NbjBDTp1xzqQ?e=024vuv)

# Issue

The "Summary of Changes" content pages in the Mauro change papers do not match up with the in-flight branches from Borland Together.

Compare the lists from the folders above and review the testing reports for further details.

The in-flight branches in question are from the NHSE TEST environment:

- CR1883
- CR1903
- CR1915

### LINKED ISSUE #72: Change papers do not show updated descriptions
# Background

This has come about due to comparisons between Borland Together change papers and Mauro generated change papers.

Originals are found in [Borland Together](https://systemccom.sharepoint.com/:f:/s/Mauro/EtIpFN2vnmhHh-EYh3N3IBoB2Da4GbdGUzEChvjiZKgpKA?e=cDxiJR)

Mauro versions are found in [Change Papers](https://systemccom.sharepoint.com/:f:/s/Mauro/Ep_lnVsG5CdIn5sCZvv6miIB_U1zbSoQ7dotYLBlxONKsA?e=anU9HS)

Test reports with more detail can be found in: [In-flight migration testing](https://systemccom.sharepoint.com/:f:/s/Mauro/Ek_bv_Mnng5ErU34rABQstsBPe8Sr6z543NbjBDTp1xzqQ?e=024vuv)

# Issue

The Mauro change papers do not show updated descriptions

# Original version

Image

# Mauro version

Image

### LINKED ISSUE #73: Change papers do not show all information for Data Sets
# Background

This has come about due to comparisons between Borland Together change papers and Mauro generated change papers.

Originals are found in [Borland Together](https://systemccom.sharepoint.com/:f:/s/Mauro/EtIpFN2vnmhHh-EYh3N3IBoB2Da4GbdGUzEChvjiZKgpKA?e=cDxiJR)

Mauro versions are found in [Change Papers](https://systemccom.sharepoint.com/:f:/s/Mauro/Ep_lnVsG5CdIn5sCZvv6miIB_U1zbSoQ7dotYLBlxONKsA?e=anU9HS)

Test reports with more detail can be found in: [In-flight migration testing](https://systemccom.sharepoint.com/:f:/s/Mauro/Ek_bv_Mnng5ErU34rABQstsBPe8Sr6z543NbjBDTp1xzqQ?e=024vuv)

# Issue

Mauro generated change papers do not show the same content as the Borland Together versions.

# Original version

Image

# Mauro version

Image

### LINKED ISSUE #74: Change papers do not show all information for Attributes
# Background

This has come about due to comparisons between Borland Together change papers and Mauro generated change papers.

Originals are found in [Borland Together](https://systemccom.sharepoint.com/:f:/s/Mauro/EtIpFN2vnmhHh-EYh3N3IBoB2Da4GbdGUzEChvjiZKgpKA?e=cDxiJR)

Mauro versions are found in [Change Papers](https://systemccom.sharepoint.com/:f:/s/Mauro/Ep_lnVsG5CdIn5sCZvv6miIB_U1zbSoQ7dotYLBlxONKsA?e=anU9HS)

Test reports with more detail can be found in: [In-flight migration testing](https://systemccom.sharepoint.com/:f:/s/Mauro/Ek_bv_Mnng5ErU34rABQstsBPe8Sr6z543NbjBDTp1xzqQ?e=024vuv)

# Issue

The Mauro generated change paper does not show all the same information for Attributes as the original Borland Together version, shows as "Also Known As" and "National Codes" tables.

# Original version

Image

# Mauro version

Image

### LINKED ISSUE #75: Change papers do not show Classes that have been changed or added
# Background

This has come about due to comparisons between Borland Together change papers and Mauro generated change papers.

Originals are found in [Borland Together](https://systemccom.sharepoint.com/:f:/s/Mauro/EtIpFN2vnmhHh-EYh3N3IBoB2Da4GbdGUzEChvjiZKgpKA?e=cDxiJR)

Mauro versions are found in [Change Papers](https://systemccom.sharepoint.com/:f:/s/Mauro/Ep_lnVsG5CdIn5sCZvv6miIB_U1zbSoQ7dotYLBlxONKsA?e=anU9HS)

Test reports with more detail can be found in: [In-flight migration testing](https://systemccom.sharepoint.com/:f:/s/Mauro/Ek_bv_Mnng5ErU34rABQstsBPe8Sr6z543NbjBDTp1xzqQ?e=024vuv)

# Issue

The Mauro change papers do not seem to ever include Classes in their change summaries.

# Original version

This was taken from the CR1915 change paper:

Image

# Mauro version

These Classes do not appear in the Mauro CR1915 change paper.

# Additional Notes

The actual issue is that the Classes are not listed as having "Changed attributes", which is why they are missing - these changes do not have a description on the class that is different. The change is therefore a list of attributes that have changed.

### LINKED ISSUE #76: Change papers for Data Set definitions refer to data set changes as "Unchanged Items"
# Background

This has come about due to comparisons between Borland Together change papers and Mauro generated change papers.

Originals are found in [Borland Together](https://systemccom.sharepoint.com/:f:/s/Mauro/EtIpFN2vnmhHh-EYh3N3IBoB2Da4GbdGUzEChvjiZKgpKA?e=cDxiJR)

Mauro versions are found in [Change Papers](https://systemccom.sharepoint.com/:f:/s/Mauro/Ep_lnVsG5CdIn5sCZvv6miIB_U1zbSoQ7dotYLBlxONKsA?e=anU9HS)

Test reports with more detail can be found in: [In-flight migration testing](https://systemccom.sharepoint.com/:f:/s/Mauro/Ek_bv_Mnng5ErU34rABQstsBPe8Sr6z543NbjBDTp1xzqQ?e=024vuv)

# Issue

When generating a "Change Paper with Data Set Definitions" from Mauro, the items originally listed as "Changed Dataset" now seem to be called "Unchanged Item". Figure out why and fix.

# Original version

This was taken from the CR1915 (Data Set) version:

Image

# Mauro version

This is the equivalent CR1915 Data Set version from Mauro:

Image

### LINKED ISSUE #77: Change paper cannot be transformed into PDF if an item description has a HTML table with changes
# Description

When testing the migration branch CR1903, there is an item which contains a HTML table in the description field. When this is turned into DITA, this causes validation errors in the OxygenXML tool because the HTML table contains the wrong number of colspans it was expecting, causing PDF transformation to fail.

# Steps to reproduce

1. Ingest May 2024 (finalised), then ingest either the CR1903 or CR1883 branches, linking them to the previous May 2024 release.
2. Run the DITA generation to produce the change paper.
3. Load the DITA into OxygenXML and try to transform into a PDF

# Expected behaviour

The DITA should be valid and the PDF should be created successfully.

# Additional notes

The cause is due to a description already existing, then containing differences in the HTML table. In the branch CR1903, there is an attribute under "Classes and Attributes > POSITION > AREA OF WORK NAME". The description field contains handwritten HTML table, where a new row was added called "Clinical Radiology":

image

If you run the Orchestrator change paper preview, you'll see what HTML it is trying to output:

image

This may be acceptable in a browser, but the OxygenXML DITA tools are far stricter with the schema:

image

### PR DESCRIPTION
**Pull request by pjmonks**
Collection of changes to support missing content in the Change Papers.

# Testing

I've tried to include some automated tests to verify the HTML/DITA output, but manual testing is slow and laborious...

1. Publish `mdm-plugin-nhs-data-dictionary` to Maven local and run in `mdm-application-build`
2. Run the Orchestrator and sign in
3. Click on the "Changes" nav link
4. Select an in-flight branch, then click "Run". This will give a preview of the change paper content.
5. Next click on "Branches" nav link.
6. Select a in-flight branch, click "Publish", then one of the generate buttons to produce DITA
7. Wait...
8. Download the zip file result, extract to a local folder
9. Use OxygenXML on the DITA content to produce a PDF

# Publication Date Format

Fixes #66 

Fixed the publication date to use "dd MMMM yyyy" format, as requested in NHSDigital/DataDictionaryPublication#498

image

# Missing descriptions for "Updated descriptions"

Fixes #72 

This didn't seem to work on TEST but does on my machine (I don't understand why...), so considering this resolved apart from styling, which is NHSE responsibility:

image

# Include "Change attributes" list for Classes

Fixes #75 

Classes did not appear in the list of changes because only descriptions were checked. I've updated this to perform a check on attributes within a class that might have been added/removed:

image

# Include missing tables for Attributes

Fixes #74 

Attributes had certain information missing, which I have now added.

Aliases table:
image

National codes table:
image

# Include "Format / Length" for Data Elements

image

# "Summary of Changes" incorrect

Resolves #71 

I think this is down to the data ingested and being modified on the server. Also, some of the supposed changes were down to missing content (above), so I think this issue is actually a red herring.

# Data Set definition changes

Resolves #76 

When including data set definitions:

1. Refer to the definitions as "Changed Data Set" instead of "Unchanged Items", like the original Borland Together change papers.
2. Don't include Data Elements and Attributes related to data sets multiple times - if a Data Element or Attribute was changed, it does not need to be added a second time, since this breaks the unique topic keys of the DITA output.

# Exclude diffs containing HTML tables

Fixes #77 

There is an issue with `DaisyDiffHelper` when comparing HTML content containing tables, the added/removed tags/markup that it adds will intermingle with the table markup, causing a table structure that is not accurate. This is especially troublesome when converting the HTML into DITA because that is much stricter on the XML schema. Temporary solution is to not perform a diff on such content - just show the current content with a message to explain.

# Include data set specifications

Resolves #73 

Added the missing data set specification tables to the change paper (preview and PDF):

image

### CODE REVIEW

File: src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDClass.groovy [hunk 1]
```diff
@@ -192,4 +196,96 @@ class NhsDDClass implements NhsDataDictionaryComponent <DataClass> {
...
+            return
+        }
+
+        Change changedAttributesChange = createChangedAttributesChange(previousClass)
+        if (changedAttributesChange) {
+            changes.add(changedAttributesChange)
+        }
+    }
+
+    Change createChangedAttributesChange(NhsDDClass previousClass) {
```
Comment (OButlerOcc) [inline src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDClass.groovy:216]: names a bit funky here but makes sense in context. Still I'd prefer CreateChangeObjectFromChangedAttributes

File: src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDCode.groovy [hunk 2]
```diff
@@ -93,4 +97,118 @@ class NhsDDCode {
...
+        Div.build {
+            simpletable(relColWidth: new SpaceSeparatedStringList (["1*", "4*"])) {
+                stHead {
+                    stentry "Code"
+                    stentry "Description"
+                }
+                currentCodes.each { code ->
+                    NhsDDCode existing = previousCodes.find { previous -> previous.code == code.code }
+                    if (!existing || (existing && existing.definition != code.definition)) {
+                        strow {
```
Comment (OButlerOcc) [inline src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDCode.groovy:162]: Not your doing, its a james thing but strow is a terrible name for anything. SimpleTableRow no?

File: src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDDataSet.groovy [hunk 3]
```diff
@@ -189,4 +188,65 @@ class NhsDDDataSet implements NhsDataDictionaryComponent <DataModel> {
...
     }
 
+    @Override
+    void buildComponentDetailsChangeList(List<Change> changes, NhsDataDictionaryComponent previousComponent, boolean includeDataSets) {
+        Change standardDescriptionChange = createStandardDescriptionChange(previousComponent, includeDataSets)
+        if (standardDescriptionChange) {
+            changes.add(standardDescriptionChange)
+        }
+
+        // Data set tables should only be included if specifically requested
```
Comment (OButlerOcc) [inline src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDDataSet.groovy:198]: I really appreciate all the comments here on the closures, makes them much easier to parse at a glance, I will make a note to start doing that in all my code

**Review by OButlerOcc — COMMENTED** [review]
I've given this a good battering and despite my fumbling it seems quite robust. I know it wouldn't fail on null changes but its good to see you've accounted for the horror of parsing xml with null checks everywhere.

I appreciate all the comments over closures, they make the processing code a lot more readable. I did find a couple of the method names weird but they make sense in context.

I noticed a situation where I was able to produce a change paper with empty changes, In this case I would like it to throw an green help box explaining there's no changes and abort rather than produce a whole empty dita project. If this is an easy change. I would suggest including it before the release. Else maybe add it as a "improvement"

I'm impressed you included unit tests for this, though they do seem a little specific they should catch any changes that bork the process.

There seems to be a bug/feature? where if you ingest a branch with no releaseDate then it gets a bit confused and the change paper produced has no changes.
It might be more of a problem with the ingest as the job looks like:
image

My ingest screen looks like:
image

Another bug? This one less serious. All my downloads are labelled CRXXXX, even for different branches. 
image

**Review by OButlerOcc — APPROVED** [review]
Accidently clicked comment instead of approve, I made the feedback in the previous review. Please check it over and if your happy with the issues I found then merge. I checked the paper against the spec and the old paper and I'm happy it does whats asked

File: src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDCode.groovy (hunk 2 above)
Comment (pjmonks) [reply src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/NhsDDCode.groovy:162]: `strow` is the tag name from the DITA schema. This is the DSL generating that tag name as output.

**Comment by pjmonks** [comment]
To answer some of your questions...

The dictionary branch should have a "Work Item" profile at the versioned folder level which details the change paper reference e.g. "CR1883". The reason why "CRXXX" is used is because we haven't overwritten that profile in our local environments. Same with the background page text, this would be included with the work item profile to produce correct page output.

I think your comments are worth noting but I'm happy to merge this now and release it for testing.

**Pull request merged**